pkg/objstore/config.go

```diff
@@ -0,0 +1,137 @@
+package objstore
+
+import (
+	"fmt"
+	"os"
+
+	"github.com/pkg/errors"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+type objProvider string
+
+const (
+	GCS         objProvider = "GCS"
+	S3          objProvider = "S3"
+	Unsupported objProvider = "UNSUPPORTED"
+)
+
+var ErrNotFound = errors.New("no valid provider configuration supplied")
+var ErrMissingGCSBucket = errors.New("missing Google Cloud Storage bucket name for stored blocks")
+var ErrInsufficientS3Info = errors.New("insufficient s3 configuration information")
+
+// BucketConfig encapsulates the necessary config values to instantiate an object store client.
+// Use Provider to represent the Object Stores
+type BucketConfig struct {
+	Provider      string
+	Bucket        string
+	Endpoint      string
+	AccessKey     string
+	secretKey     string
+	Insecure      bool
+	SignatureV2   bool
+	SSEEncryption bool
+}
+
+// NewBucketConfig return the configuration of object store
+// TODO(jojohappy) should it support multiple bucket?
+func NewBucketConfig(cmd *kingpin.CmdClause) *BucketConfig {
+	var bucketConfig BucketConfig
+
+	cmd.Flag("provider.type", "Specify the provider for object store. If empty or unsupport provider, Thanos won't read and store any block to the object store. Now supported GCS / S3.").
+		PlaceHolder("<provider>").StringVar(&bucketConfig.Provider)
+
+	cmd.Flag("provider.bucket", "The bucket name for stored blocks.").
+		PlaceHolder("<bucket>").Envar("PROVIDER_BUCKET").StringVar(&bucketConfig.Bucket)
+
+	cmd.Flag("provider.endpoint", "The object store API endpoint for stored blocks. Support S3-Compatible API").
+		PlaceHolder("<api-url>").Envar("PROVIDER_ENDPOINT").StringVar(&bucketConfig.Endpoint)
+
+	cmd.Flag("provider.access-key", "Access key for an object store API. Support S3-Compatible API").
+		PlaceHolder("<key>").Envar("PROVIDER_ACCESS_KEY").StringVar(&bucketConfig.AccessKey)
+
+	// TODO(jojohappy): should it be encrypted?
```

Comment:
No, users should provide the original password or key, but the `BucketConfig` will encrypt that and return it decrypted. Because the `s3.Config` did not export the `secretKey`, I didn't know why did that. In the new `BucketConfig`, I still doesn't export this field, so I think should it be encrypted? 

I am confusing that, if it is not necessary, I will remove that.